### PR TITLE
feat(bzlmod): support bazel downloader when downloading wheels

### PR DESCRIPTION
**Pull request by aignas**
This introduces 3 attributes and the minimal code to be able to download wheels
using the bazel downloader for the host platform. This is not yet adding
support for targeting a different platform but just allows us to get the wheels
for the host platform instead of using `pip`.

All of this is achieved by calling the PyPI's SimpleAPI (Artifactory should work
as well) and getting the all URLs for packages from there. Then we use the `sha256`
information within the requirements files to match the entries found on SimpleAPI
and then pass the `url`, `sha256` and the `filename` to `whl_library`, which uses
`repository_ctx.download`.

If we cannot find any suitable artifact to use, we fallback to legacy `pip` behaviour.

Testing notes:
* Most of the code has unit tests, but the `pypi_index.bzl` extension could have more.
* You can see the lock file for what the output of all of this code would be on your
  platform.
* Thanks to @dougthor42 for testing this using the credentials helper against a private
  registry that needs authentication to be accessed.

Work towards #1357

### CODE REVIEW

File: python/private/bzlmod/BUILD.bazel
Comment (Wyverald) [inline python/private/bzlmod/BUILD.bazel:54]: as a stopgap, you could list the .bzl files here explicitly (before https://github.com/bazelbuild/platforms/pull/89) is discussed and merged

File: tests/private/whl_target_platforms/select_whl_tests.bzl
Comment (rickeylev) [inline tests/private/whl_target_platforms/select_whl_tests.bzl:1]:
```suggestion
# Copyright 2024 The Bazel Authors. All rights reserved.
```

File: python/private/bzlmod/BUILD.bazel
Comment (aignas) [reply python/private/bzlmod/BUILD.bazel:54]: @Wyverald, if I understand correctly, I can't do it. This is because the host platform extension is internal to the platforms module. The extension that creates it is used internally by platforms and I am not sure I should use the extession myself in rules python. Or maybe it is exactly how I should consume this code? Please advise. :)

File: internal_deps.bzl
Comment (aignas) [reply internal_deps.bzl:66]: This was duplicated, it was present twice in the same file.

File: python/private/bzlmod/BUILD.bazel
Comment (aignas) [reply python/private/bzlmod/BUILD.bazel:54]: Anyway, I decided to not use `platforms` in the repository context, the PRs are now decoupled, we can discuss it in bazelbuild/platforms#89.

File: python/private/bzlmod/pip.bzl
Comment (aignas) [reply python/private/bzlmod/pip.bzl:284]: Added a comment on why I am doing this, let me know what you think.

File: python/private/bzlmod/pip.bzl
Comment (aignas) [reply python/private/bzlmod/pip.bzl:459]: There are two things in `pip`:
* `--index-url` - this is the default index
* `--extra-index-urls` - these are the extra indexes that would be used in addition to the one above. I am not supporting this usecase as it makes the code more complex.

I see how the `ex` word here may make people think about `--extra-index-urls` argument. Shall I add something in the docs?

File: python/private/bzlmod/pip.bzl
Comment (aignas) [reply python/private/bzlmod/pip.bzl:465]: I thought I have removed the TODO comments.

File: python/private/pypi_index.bzl
Comment (aignas) [reply python/private/pypi_index.bzl:173]: Yes, I was thinking that non-simple api sources (like regular downloads) can be also added later. Should we make it a struct later?

File: python/private/pypi_index.bzl
Comment (aignas) [reply python/private/pypi_index.bzl:182]: This is a slightly different kind of parsing and we need to somehow separate/extract the `sha256` values. I could modify the original parser, but did not want to update that code in case this has bugs. Maybe I could add a note here to refactor later once we stabilize all of this?

File: python/private/pypi_index.bzl
Comment (aignas) [reply python/private/pypi_index.bzl:206]: Changed to just `requirement`, which is vague, but it might be good enough in the context of the struct and it makes it clearer to understand where it is used.

File: python/private/pypi_index.bzl
Comment (aignas) [reply python/private/pypi_index.bzl:256]: I think the `yanked` status is more important for SAT solvers which need to traverse the possible versions. They should never add a yanked version in to a lock file. This failure behaviour is a way for us to signal that the version was yanked due to some reason as we are not doing relocking every time. I am not sure if `pip` would refuse installing a yanked package though.

Would you prefer me to check? I think telling the user to relock is a good message, but at the same time if the shas are in the `MODULE.bazel.lock` file this error would happen only during initial extension evaluation time.

Comment (aignas) [reply python/private/pypi_index.bzl:256]: Reworked it and now we are just ignoring it and if we will not get any wheels because of this, `pip` will do whatever it normally does.

File: python/private/pypi_index.bzl
Comment (aignas) [reply python/private/pypi_index.bzl:84]: Added a comment to explain.

File: python/private/bzlmod/pip.bzl
Comment (aignas) [reply python/private/bzlmod/pip.bzl:203]: I'll think how to refactor that. I think it may be probably better to wait until we are fetching wheels for all platforms as that will cause some churn of the code as well.

File: python/private/bzlmod/pip.bzl
Comment (dougthor42) [reply python/private/bzlmod/pip.bzl:461]: IME it's very rare to have a full private index[^1]. Instead, the common case is to use the public PyPI.org as one index and then add the private index via `--extra-index-urls`.

So I think _only_ supporting `--index-url` will not help many people :worried:.

[^1]: Though from a security perspective this is the best thing to do: have a mirror of public pypi where you allowlist only the packages that people need and denylist everything else. The problem is that's a lot of overhead and it's seen as "not worth it" :disappointed:

Comment (aignas) [reply python/private/bzlmod/pip.bzl:461]: If we add an extra index it would be good to know what packages we should fetch from there because we would get 404 errors at best in the pypi public and wrong results at worst.

For artifactory, you can create virtual repos that would be a union of pypi index and the internal one. The logic of what comes from which index would be handled by the index. I'm not sure about other indexes though.

I am thinking that even though this might not be the best API now, I don't see how we could get it better with what we have now. Maybe in the long run we could use uv/pip install dry-run report to get an idea about which package comes from which index, but I would love to postpone this to a future PR. @philsc has done greate work to get the install reports integrated in rules_python, so maybe we can collaborate to come up with something better some time later? :)

File: MODULE.bazel
Comment (aignas) [reply MODULE.bazel]: Removed the need for that by copying two functions.

File: .bazelversion
Comment (aignas) [reply .bazelversion]: Reverted.

File: python/private/bzlmod/pip.bzl
Comment (dougthor42) [reply python/private/bzlmod/pip.bzl:474]: ❤️ thank you!

I'll say that getting 404s in public pypi is already how `pip` works with `--extra-index-url` so I'm not sure if that's a dealbreaker. `pip` first tries to find the package at `--index-url` and if it fails there it goes to `--extra-index-url`s in order. (This is a Bad Thing :tm: because of exactly what you mentioned[^1], but sadly that's how it currently works.)

> For artifactory, you can create virtual repos that would be a union of pypi index and the internal one.

Ooooo I didn't know that, that's a nice feature! Looks like Google Artifact Registry has that too - I'll have to tell our SWOps people.

[^1]: Example: Malicious actor finds out that we have a package `foo-bar-company` on our private index and then namesquats a bad package with the same name on the public index. 👎

File: CHANGELOG.md
Comment (dougthor42) [inline CHANGELOG.md:55]: `experimental_extra_index_urls` too.

**Comment by dougthor42** [comment]
This is supposed to be able to use a credential helper, correct? It looks like it does - I can force various errors by having either invalid JSON or incorrect creds in the helper.

The issue is that I get 404s for packages that should exist in the private index **if the package contains a dash `-` character**.

Here's what I mean. Our private index has two packages in it:

+ `company-code-package`
+ `ourpackage`

If my `requirements_lock.txt` has only `ourpackage` in it, I can `bazel build` just fine. Things use the bazel downloader and everything's rainbows.

If it has `company-code-package` in it, we get a 404:

```
$ bazel build //:foo
WARNING: Download from https://pypi.org/simple/ourpackage/ failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://pypi.org/simple/company_code_package/ failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://[REDACTED]/simple/attrs/ failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://[REDACTED]/simple/pandas/ failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://[REDACTED]/simple/numpy/ failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://[REDACTED]/simple/company_code_package/ failed: class java.io.FileNotFoundException GET returned 404 Not Found
ERROR: Traceback (most recent call last):
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/bzlmod/pip.bzl", line 447, column 30, in _pip_impl
                _create_whl_repos(module_ctx, pip_attr, hub_whl_map, whl_overrides, simpleapi_cache)
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/bzlmod/pip.bzl", line 186, column 40, in _create_whl_repos
                index_urls = simpleapi_download(
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/pypi_index.bzl", line 125, column 17, in simpleapi_download
                fail("Failed to download metadata about '{}' from urls: {}".format(
Error in fail: Failed to download metadata about 'company_code_package' from urls: https://pypi.org/simple, https://[REDACTED]/simple
ERROR: error evaluating module extension pip in @@rules_python~//python/extensions:pip.bzl
INFO: Elapsed time: 1.284s
INFO: 0 processes.
ERROR: Build did NOT complete successfully
FAILED:
    Fetching module extension pip in @@rules_python~//python/extensions:pip.bzl; starting
```

Things to note:

+ the 404's for `https://[REDACTED]/simple/attrs`, `numpy`, and others are expected - those are available on public pypi.
+ the 404 for `https://pypi.org/simple/ourpackage/` is expected - that package is only available on our private index.
+ the url `https://[REDACTED]/simple/company_code_package/` uses underscores instead of dashes.

That third point is the kicker. When a normal `pip install company-code-package` is done, the GET request uses the hyphenated name but the response returns links with the underscored name.

```
$ pip install company-code-package -vvvv
...
Request header has "max_age" as 0, cache bypassed
https://us-west2-python.pkg.dev:443 "GET /qh-lab/qh-py/simple/company-code-package/ HTTP/1.1" 200 None
Updating cache with response from "https://[REDACTED]/simple/company-code-package/"
  Found link https://[REDACTED]/company-code-package/company_code_package-1.0-py3-none-any.whl#sha256=7aaf9af436d61e73e3c486e5dc7ca7563f51b679f7c34bbe14540f7a9e18b5dd (from https://[REDACTED]/simple/company-code-package/), version: 1.0
...
```

Could it be that the python package name sanitation is happening too early?

**Comment by aignas** [comment]
I was normalizing the names in the requirements to use underscores. That works with PyPI, but it seems that it does not work with every index. I have changed the code to use exactly the same package name as is in the supplied requirements file.

I will be adding various unit tests from now on, so expect a little bit more activity here.

Open questions:
* Should we just have a single array for the `index_urls`?
* What kind of docs should we have for documenting that we support credential helper. @dougthor42 , maybe you could help me with a paragraph on how you got it working? (Assuming the normalization was the actual issue)

**Comment by dougthor42** [comment]
> What kind of docs should we have for documenting that we support credential helper. @dougthor42 , maybe you could help me with a paragraph on how you got it working?

I'd be more than happy to! I think I'll be able to get something whipped up within a day or three. I'll just submit a PR onto `aignas:feat/pip-simpleapi`.


> I have changed the code to use exactly the same package name as is in the supplied requirements file.

Thanks! That part works wonderfully :grin:.

### Other Issues/Comments

1. I'm now having issues because some of the ~440 packages we use _don't have wheels_ but that's **out of scope of this PR**.
2. I get spammed with 404 warnings when the downloader tries to find things like `numpy` in our private index.
     + Perhaps it can be downgraded to INFO or DEBUG? That said, **it's not a big issue and we can resolve it in other ways** (like the virtual index mentioned elsewhere).
3. ValueErrors in `whl_target_platforms.bzl`. Some packages have nonstandard / invalid / just old platform descriptors.
    + [grpcio v1.62.1](https://pypi.org/project/grpcio/1.62.1/#files) has `linux_armv7l`
    + [freetype_py v2.1.0](https://pypi.org/project/freetype-py/2.1.0/#files) has `macosx_10_6_intel`
    + [libclang v13.0.0](https://pypi.org/project/libclang/13.0.0/#files) has `manylinux_2_17_armv7l`
    + [watchdog v3.0.0](https://pypi.org/project/watchdog/3.0.0/#files) has `win_ia64`
    + Those are all the ones I found in our project.
    + I had to add a debug statement to `parse_whl_name.bzl` so that it would print the package that was being processed when the error occured - it might be nice to pass `file` down to `whl_target_platforms._cpu_from_tag` so that the error message has the package name in it.


Here's an example of the error in (3):

```
ERROR: Traceback (most recent call last):
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/bzlmod/pip.bzl", line 447, column 30, in _pip_impl
                _create_whl_repos(module_ctx, pip_attr, hub_whl_map, whl_overrides, simpleapi_cache)
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/bzlmod/pip.bzl", line 249, column 29, in _create_whl_repos
                whl = select_whl(
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/whl_target_platforms.bzl", line 199, column 41, in select_whl
                platforms = whl_target_platforms(platform_tag)
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/whl_target_platforms.bzl", line 234, column 25, in whl_target_platforms
                cpus = _cpu_from_tag(platform_tag)
        File "/usr/local/google/home/dthor/.cache/bazel/_bazel_dthor/dbe74c4144b5c9a438d84a119652bef9/external/rules_python~/python/private/whl_target_platforms.bzl", line 268, column 13, in _cpu_from_tag
                fail("Unrecognized tag: '{}': cannot determine CPU".format(tag))
Error in fail: Unrecognized tag: 'macosx_10_6_intel': cannot determine CPU
ERROR: error evaluating module extension pip in @@rules_python~//python/extensions:pip.bzl
```

File: python/private/pypi_index.bzl
Comment (rickeylev) [reply python/private/pypi_index.bzl]: I think a struct is fine given the plans

File: python/private/pypi_index.bzl
Comment (rickeylev) [reply python/private/pypi_index.bzl:242]: Ah ok. I think its fine then.

File: python/private/bzlmod/pip.bzl
Comment (rickeylev) [reply python/private/bzlmod/pip.bzl:487]: I think this convo can be resolved? I'll mark it as such

File: python/private/bzlmod/pip.bzl
Comment (rickeylev) [reply python/private/bzlmod/pip.bzl:306]: Ahhh, thanks. Yeah that clarifies. Suggested an edit to drop the "maybe it won't be needed in the future". If there's a particular feature request somewhere to reference refer to that.

File: python/private/bzlmod/pip.bzl
Comment (rickeylev) [inline python/private/bzlmod/pip.bzl:305]:
```suggestion
        # args are manipulated in the code going before.
```

File: python/private/bzlmod/pip.bzl
Comment (rickeylev) [reply python/private/bzlmod/pip.bzl:204]: Fair, SGTM

**Review by rickeylev — APPROVED** [review]
LGTM. It sounds like dougthor42 identified a couple edge cases, but I think the main part of this is good now, so approving so you can later merge

**Comment by dougthor42** [comment]
After 5643866 it looks like everything's working very well! 404s are still spammy but I can live with that. From what I can tell, both wheels and sdists were correctly installed, it didn't fail out on unknown platforms, and it even fell back to standard pip for one of our more annoying packages [klayout](https://pypi.org/project/klayout).

This is so awesome, thank you very much!

**Comment by aignas** [comment]
I'll merge this and start looking at supporting multi-platform whl selects which is what #1744 tried to achieve.

**Comment by aignas** [comment]
This is the first time I see a non closed PR after merging. I'll just close this one manually.

File: MODULE.bazel
Comment (hunshcn) [inline MODULE.bazel:61]: Is it intentional not to use PIP_INDEX_URL env here, and why?

Comment (aignas) [reply MODULE.bazel:61]: If you want to override this value you can use bazel downloader config. In the general sense it is not a given that people will have their `PIP_INDEX_URL` configured to contain the index that works for fetching these dependencies.

Hence to avoid unintended breakage, this is hardcoded.

**Pull request closed without merging**
